packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+      const dialogElement = dialogRef.current;
+      if (!dialogElement) {
+        return;
+      }
+
+      if (!hasNativeDialog && isModal) {
+        (dialogElement.nextSibling as HTMLDivElement).classList.remove(
+          classes['backdrop-visible'],
+        );
+      }
+      onCloseStart?.();
+      // trigger closing of the dialog after animation
+      setTimeout(() => {
+        if (dialogElement.open) {
+          dialogElement.close();
+        }
+      }, animationDuration);
+    }, [hasNativeDialog, animationDuration, isModal, onCloseStart]);
+
+    useClickOutside([dialogRef], () => handleDialogClose(), open && !isModal);
+    useEscapeKey(() => handleDialogClose(), open && !isModal);
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+
+      if (!dialogElement) {
+        return undefined;
+      }
+
+      if (!open) {
+        dialogElement.returnValue = '';
+      }
+      if (open) {
+        if (!dialogElement.open) {
+          if (isModal) {
+            dialogElement.showModal();
+          } else {
+            dialogElement.show();
+          }
+
+          if (!hasNativeDialog && dialogElement.nextSibling) {
+            // use the polyfill backdrop
+            (dialogElement.nextSibling as HTMLDivElement).classList.add(
+              classes['backdrop-visible'],
+              classes.backdrop,
+            );
+            // intercept and prevent modal closing if preventClose is true
+            (dialogElement.nextSibling as HTMLDivElement).addEventListener(
+              'click',
+              onPolyfillBackdropClick,
+            );
+          }
+        }
+      } else if (dialogElement.open) {
+        handleDialogClose();
+      }
+
+      return () => {
+        if (dialogElement.open) {
+          dialogElement.close('skipOnClose');
+        }
+
+        if (!hasNativeDialog && dialogElement.nextSibling) {
+          (dialogElement.nextSibling as HTMLDivElement).removeEventListener(
+            'click',
+            onPolyfillBackdropClick,
+          );
+        }
+      };
+    }, [
+      isModal,
+      open,
+      handleDialogClose,
+      hasNativeDialog,
+      onPolyfillBackdropClick,
+    ]);
+
+    const onDialogClick = (
+      event: ClickEvent<HTMLDialogElement> | ClickEvent<HTMLDivElement>,
+    ) => {
+      // the dialog content covers the whole dialog element
+      // leaving the backdrop element as the only clickable area
+      // that can trigger an onClick event
+      if (event.target === event.currentTarget && !preventClose) {
+        handleDialogClose();
+      }
+    };
+
+    return (
+      <>
+        {/* eslint-disable-next-line  jsx-a11y/no-noninteractive-element-interactions */}
+        <dialog
+          onClick={onDialogClick}
+          className={clsx(classes.base, isModal && classes.modal, className)}
+          ref={applyMultipleRefs(ref, dialogRef)}
+          {...rest}
+        >
+          {isModal && (
```

Comment:
Should the close button be shown when `preventClose` is set? 

```suggestion
          {isModal && !preventClose && (
```

Reply:
Hmm, good question. So far I've used `preventClose` to disable closing the dialog when the escape key is pressed or when the user clicks outside the dialog, to avoid "accidental" closing for critical tasks. So far I was convinced that a modal dialog should _always_ have a clear, available way of closing for the sake of accessibility (What if developers don't provide that in the content transmitted to the dialog ? ). WDYT ?

Reply:
As a developer, I would expect any default mechanism of closing the dialog to be prevented. In other words, I can see developers omitting the `onClose` callback, assuming that it won't be called.

> a modal dialog should always have a clear, available way of closing for the sake of accessibility

I agree! This becomes the responsibility of the developers when setting `preventClose`. 

I can see this prop being useful for [alert](https://developer.mozilla.org/en-US/docs/Web/API/Window/alert) or [prompt](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt) style dialogs. Having a close button could be confusing for these use cases.

Image a user tries to close a draft invoice and this dialog pops up:

> Do you want to save this invoice?
> 
> [Cancel] [Save]

What would/should happen when a user presses the close button? Save or cancel? Because of this ambiguity, I think it's best to remove it.